src/goto-instrument/contracts/instrument_spec_assigns.cpp

```diff
@@ -91,30 +94,119 @@ void instrument_spec_assignst::check_inclusion_assignment(
+    "the instrumented function must exist in the model");
+
+  const goto_programt &body = found->second.body;
+  traverse_instructions(
+    body.instructions.begin(), body.instructions.end(), covered_lines);
+
+  std::unordered_set<symbol_exprt, irep_hash> symbols;
+  collect_static_symbols(covered_lines, symbols);
 
-  for(const auto &sym_pair : st)
+  for(const auto &sym : symbols)
+    create_snapshot(create_car_from_stack_alloc(sym), dest);
+}
+
+size_t line_to_size_t(const irep_idt &line)
+{
+  PRECONDITION(!line.empty());
+  std::stringstream stream(id2string(line));
+  size_t res;
+  stream >> res;
+  return res;
+}
+
+void instrument_spec_assignst::traverse_instructions(
+  const goto_programt::const_targett &begin,
+  const goto_programt::const_targett &end,
+  covered_linest &covered_lines) const
+{
+  goto_programt::const_targett it(begin);
+
+  while(it != end)
   {
-    const auto &sym = sym_pair.second;
-    if(sym.is_static_lifetime)
+    const auto &loc = it->source_location();
+    const auto &loc_fun = loc.get_function();
+    if(!loc_fun.empty())
     {
-      auto fname = sym.location.get_function();
-      if(
-        !fname.empty() &&
-        (fname == function_id || call_graph.get_node_index(fname).has_value()))
+      auto &itv = covered_lines[loc_fun];
+      const auto &loc_line = loc.get_line();
+      if(!loc_line.empty())
+        itv.update(line_to_size_t(loc_line));
+    }
+
+    if(it->is_function_call())
+    {
+      const auto &fun_expr = it->call_function();
+
+      PRECONDITION_WITH_DIAGNOSTICS(
+        fun_expr.id() == ID_symbol,
+        "local static search requires function pointer removal");
+      const irep_idt &fun_id = to_symbol_expr(fun_expr).get_identifier();
+
+      const auto &found = functions.function_map.find(fun_id);
+
+      if(found != functions.function_map.end())
+      {
+        const goto_programt &body = found->second.body;
+        if(covered_lines.find(fun_id) == covered_lines.end())
+        {
+          traverse_instructions(
+            body.instructions.begin(), body.instructions.end(), covered_lines);
+        }
+      }
+      else
+      {
+        covered_lines[fun_id];
+      }
+    }
+    it++;
+  }
+}
+
+void instrument_spec_assignst::collect_static_symbols(
+  covered_linest &covered_lines,
+  std::unordered_set<symbol_exprt, irep_hash> &dest)
+{
+  while(true)
+  {
+    bool fixpoint = true;
+    for(const auto &sym_pair : st)
+    {
+      const auto &sym = sym_pair.second;
+      if(sym.is_static_lifetime)
       {
-        // We found a symbol with
-        // - a static lifetime
-        // - non empty location function
-        // - this function appears in the call graph of the function
-
-        // insert in tracking set and generate snapshot instructions
-        // for this target.
-        create_snapshot(create_car_from_stack_alloc(sym.symbol_expr()), dest);
+        const auto &loc_fun = sym.location.get_function();
+        if(!loc_fun.empty())
```

Comment:
As a second layer of defense: have you also consider evaluating the symbol name to extract a function prefix from it?